Implement #1611 - provide dynamic debug config

rkeithhill opened the pull request:
## PR Summary

This an overdue, first-pass attempt at implementing dynamic debug configurations as suggested by @isidorn in #1611.  I'm completely open to changing the text that appears in the various prompts.  I'm also open to the idea that "Attach to PowerShell Host Process" is rare enough that we could eliminate one level of prompting and only allow dynamic debug config for "Launch".  Or maybe we make the second level quick pick a multi-select option and then create the debug configs the user selects?  Then we could remove the first quick pick prompt.

BTW this functionality **only** comes into play when you don't have a launch.json file. The first time you configure launch.json is where you will see this.  Once launch.json exists, you are always choosing from the list of debug config snippets we provide in our package.json file.  One benefit of this approach is that when a user goes through this process now, they get just one debug config added to their new launch.json as opposed to everything we used to define in our package.json's initialDebugConfigurations - which was a lot of stuff.  

@isidorn I tried all the other debuggers on my system in this mode and none of them prompt (node, chrome, C++ gdb/Windows, .NET Core).  Are you sure this is OK for PowerShell to do?  Also, would you mind reviewing this PR?

Here is what this process looks like:

![DebugLaunchConfig1](https://user-images.githubusercontent.com/5177512/61177955-30493d00-a59f-11e9-8969-f57f84ca0d4b.gif)

![DebugLaunchConfig2](https://user-images.githubusercontent.com/5177512/61177957-33442d80-a59f-11e9-80d5-df8ac4209679.gif)

![DebugLaunchConfig3](https://user-images.githubusercontent.com/5177512/61177958-35a68780-a59f-11e9-9bbb-2f7071645423.gif)

![DebugAttachConfig1](https://user-images.githubusercontent.com/5177512/61177959-38a17800-a59f-11e9-9033-5add41730b44.gif)

Finally, I removed the `temp integrated console debug` config snippet.  That will still continue to work but I think we should remove this for now as its a bug factory.  Also, looking at this list of snippets, I wonder if we could cull it further:

![image](https://user-images.githubusercontent.com/5177512/61177974-80280400-a59f-11e9-9fe4-fea5a9923eae.png)

I'm fuzzy on what the difference is between `PowerShell: Interactive Session` and `PowerShell: Attach Interactive Session`.  Also, not sure the `w/Args` is worth having as a snippet. I did add to the docs on `args` that you could specify `${command:SpecifyScriptArgs}` to be prompted for args.  Thoughts?

Also, one last thing.  We had the breakpoints configuration in a part of package.json that isn't valid anymore / deprecated.  So, I moved it to where the VSCode debug docs say it should go.  That said, I'm not sure when that switchover happened.

## PR Checklist

Note: Tick the boxes below that apply to this pull request by putting an `x` between the square brackets.
Please mark anything not applicable to this PR `NA`.

- [x] PR has a meaningful title
- [x] Summarized changes
- [ ] PR has tests
- [ ] This PR is ready to merge and is not work in progress
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready

rjmholt commented on src/features/DebugSession.ts:
```diff
@@ -42,6 +43,88 @@ export class DebugSessionFeature implements IFeature, DebugConfigurationProvider
+
+        let debugConfiguration = [];
+
+        if (debugTypeSelection.label === "Launch") {
```
This logic here I would write as:

```typescript
const powerShellType = 'PowerShell';

if (debugTypeSelection.label !== "Launch") {
    return [
        {
            name: 'PowerShell: Attach to PowerShell Host Process',
            type: powerShellType,
            request: 'attach',
            runspaceId: 1
        }
    ]
}

const launchCurrentFileLabel = 'Launch Current File';
const launchScriptLabel = 'Launch Script';
const interactiveSessionLabel = 'Interactive Session';

const launchItems = [
    {
        label: launchCurrentFileLabel,
        description: 'Debugs the script currently in the active editor window'
    },
    {
        label: launchScriptLabel,
        description: 'Debugs the specified script'
    },
    {
        label: interactiveSessionLabel,
        description: 'Debugs scripts or modules executed from the Integrated Console'
    }
];

const launchSelection = await vscode.window.showQuickPick(
    launchItems,
    { placeHolder: 'Select a debugger launch option'});

if (launchSelection.label === launchCurrentFileLabel) {
    return [
        {
            name: 'PowerShell: Launch Current File',
            type: powerShellType,
            request: 'launch',
            script: '${file}',
            cwd: '${file}'
        }
    ]
}

if (launchSelection.label === launchScriptLabel) {
    return [
        {
            name: 'Powershell: Launch Script',
            type: powerShellType,
            request: 'launch',
            script: 'Enter path or command to execute, e.g.: ${workspaceFolder}/src/foo.ps1 or Invoke-Pester'
        }
    ]
}

// Last remaining possibility is interactive launch
return [
    {
        name: 'PowerShell: Interactive Session',
        type: powerShellType,
        request: 'launch',
        cwd: ''
    }
];
```

rkeithhill replied:
If the impl remains  the way it is then yeah, it should be broken up into mutiple methods like you show above.  At this point, I'm more interested in feedback on the experience.  I think two levels of prompts is probably one too many.  Yesterday, I was thinking that perhaps the multi-selection approach would be better.  I'm really interested in getting @isidorn feedback because none of the other debuggers except for well, Python. It does prompt with this.  

![image](https://user-images.githubusercontent.com/5177512/61233950-0b63df80-a6ef-11e9-90b7-7fbc5882a23a.png)

BTW with that Python debug prompt above, if I press <kbd>escape</kbd> to cancel the operation, I get all the Python configs dumped into my launch.json.   That seems like a bug.

rkeithhill commented:
This last commit reduces the number of prompts to one:

![image](https://user-images.githubusercontent.com/5177512/61585103-4613b080-ab11-11e9-954e-e78d192dc7a0.png)

Still open to wording changes.  I updated the wording of the `Launch Script` debug configuration snippet to show how you can invoke arbitrary commands like `Invoke-Pester` e.g.:
```json
        {
            "name": "PowerShell: Launch Script",
            "type": "PowerShell",
            "request": "launch",
            "script": "enter path or command to execute e.g.: ${workspaceFolder}/src/foo.ps1 or Invoke-Pester",
            "cwd": "${workspaceFolder}"
        }
```
Given the above **and** the fact that you can use code lens to debug Pester tests (the most common way to debug Pester tests IMO), I've removed the `Launch Pester Tests` debug config snippet leaving these:

![image](https://user-images.githubusercontent.com/5177512/61585127-e964c580-ab11-11e9-800e-72164f4e01af.png)

I'd even like to get rid of the `w/args` snippet.  When a user gets intellisense on the `args` parameter they see how they can make their debug configuration prompt for args:

![image](https://user-images.githubusercontent.com/5177512/61585138-2af57080-ab12-11e9-8eaa-78f1fce88247.png)

Plus this prompt functionality applies to both `Launch Current File` as well as `Launch Script` so it's a bit weird only having this snippet for one and not the other.  Thoughts?

TylerLeonhardt commented on src/features/DebugSession.ts:
```diff
@@ -42,6 +43,80 @@ export class DebugSessionFeature implements IFeature, DebugConfigurationProvider
+            },
+        ];
+
+        const launchSelection =
```
Still can't get the index of the selection, huh?

rkeithhill replied:
Nah, I think you get back the selected item object.  I could add an index/id to the item objects and then test against that.  Worth it?

TylerLeonhardt replied:
Can you try it and let us know? I'm also curious for other situations as well.

rkeithhill replied:
I looked at the API, it returns `Thenable<T | undefined>` where `T` is the type of the items array provided as input:

![image](https://user-images.githubusercontent.com/5177512/61649799-635e9100-ac6f-11e9-82b0-bf4548046ebc.png)

TylerLeonhardt replied:
Ah ok so in theory if our type had an index property, we could take advantage of it

rkeithhill commented:
That last commit whittles the debug configuration snippets to just these:

![image](https://user-images.githubusercontent.com/5177512/61599810-952d1480-abe9-11e9-9ed7-284ee03569c0.png)

rkeithhill commented:
@rjmholt Are you OK with the latest changes?

rjmholt commented on src/features/DebugSession.ts:
```diff
@@ -2,9 +2,10 @@
  * Copyright (C) Microsoft Corporation. All rights reserved.
  *--------------------------------------------------------*/
 
+import path = require("path");
```
```suggestion
import * as path from "path";
```

rkeithhill replied:
Removed this since we weren't using `path` anywhere in the file.

rjmholt commented on src/features/DebugSession.ts:
```diff
@@ -42,6 +43,85 @@ export class DebugSessionFeature implements IFeature, DebugConfigurationProvider
+                    cwd: "${file}",
+                },
+            ];
+        } else if (launchSelection.id === launchScriptId) {
```
```suggestion
        }
        
        if (launchSelection.id === launchScriptId) {
```

rjmholt commented on src/features/DebugSession.ts:
```diff
@@ -42,6 +43,85 @@ export class DebugSessionFeature implements IFeature, DebugConfigurationProvider
+                    cwd: "${workspaceFolder}",
+                },
+            ];
+        } else if (launchSelection.id === interactiveSessionId) {
```
```suggestion
        }
        
        if (launchSelection.id === interactiveSessionId) {
```

isidorn commented:
@rkeithhill I appologise for not reviewing - I was out on vacation for 3 weeks.
Though I see this is merged and it looks good.
If you roll it out in insiders you should get immediate user feedback to potentially further polish this.

Thanks!

rkeithhill commented:
@isidorn No worries. I'm pretty happy with where we ended up.  And we can always course-correct based on user feedback.
